akka-sample-kafka-to-sharding-scala/processor/src/main/scala/akka/kafka/KafkaClusterSharding.scala

```diff
@@ -0,0 +1,145 @@
+      partition.toString
+    }
+  }
+
+  final class KafkaShardingMessageExtractor[M](val kafkaPartitions: Int)
+    extends ShardingMessageExtractor[ShardingEnvelope[M], M] with KafkaClusterSharding {
+    override def entityId(envelope: ShardingEnvelope[M]): String = envelope.entityId
+    override def unwrapMessage(envelope: ShardingEnvelope[M]): M = envelope.message
+  }
+
+  final class KafkaShardingNoEnvelopeExtractor[M](val kafkaPartitions: Int, entityIdExtractor: M => String)
+    extends ShardingMessageExtractor[M, M] with KafkaClusterSharding {
+    override def entityId(message: M): String = entityIdExtractor(message)
+    override def unwrapMessage(message: M): M = message
+  }
+
+  // TODO:
+  // - will require `akka-actors-typed` as another provided dep, or should we just return a classic actor?
+  // - returning a typed actor is more flexible for the user so that they can easy create it under the `user` guardian
+  //   when running akka typed. an alternative would be to create the actor ourself as a system actor, like is done with
+  //   the KafkaConsumerActor for the metadata client.
+  /**
+   * The [[RebalanceListener]] handles [[TopicPartitionsAssigned]] events created by the Kafka consumer group rebalance
+   * listener. As partitions are assigned to this consumer group member we update the External Sharding strategy
+   * accordingly so that entities are (eventually) routed to the local Akka cluster member.
+   */
+  object RebalanceListener {
+    def apply(typeKey: EntityTypeKey[_]): Behavior[ConsumerRebalanceEvent] =
+      Behaviors.setup { ctx =>
+        val typeKeyName = typeKey.name
+        val shardAllocationClient = ExternalShardAllocation(ctx.system).clientFor(typeKeyName)
+        val address = Cluster(ctx.system).selfMember.address
+        Behaviors.receive[ConsumerRebalanceEvent] {
+          case (ctx, TopicPartitionsAssigned(_, partitions)) =>
+            import ctx.executionContext
+            val partitionsList = partitions.mkString(",")
+            ctx.log.info("Consumer group '{}' is assigning topic partitions to cluster member '{}': [{}]",
+              typeKeyName, address, partitionsList)
+            val updates = partitions.map { tp =>
+              val shardId = tp.partition().toString
+              // Kafka partition number becomes the akka shard id
+              // TODO: support assigning more than 1 shard id at once?
+              shardAllocationClient.updateShardLocation(shardId, address)
+            }
+            // TODO: pipeToSelf since we're closing over local state?
+            Future
+              .sequence(updates)
+              // each Future returns successfully once a majority of cluster nodes receive the update.
+              // there's no point blocking here because the rebalance listener is triggered asynchronously.  if we want
+              // to block rebalances then we should provide an implementing using the `PartitionAssignmentHandler` instead
```

Comment:
Usually Kafka consumer group rebalance handlers will let the user block while the rebalance is happening, so that they can do some action. For example, flush latest commits when partitions are revoked before another group member starts reading from that partition.  However, the `ConsumerRebalanceEvent` implementation in Alpakka Kafka is asynchronous and does not allow for blocking.  If we think the blocking would be useful we could instead implement the `PartitionAssignmentHandler` as noted in the comment above.